doc/developer/design/20240609_error_handling.md

```diff
@@ -0,0 +1,416 @@
+ `value`        | `bytea`        | Yes      | The value bytes of the Kafka message that failed to decode.
+
+The DLQ table works like a subsource. It can be queried via `SELECT`, subscribed
+to via `SUBSCRIBE`, and used as the upstream relation for a Kafka sink.
+
+The DLQ table is append only. Allowing users to specify a retention policy on
+the DLQ table is left to future work. Users cannot issue `INSERT`, `UPDATE`, or
+`DELETE` commands against the DLQ[^2].
+
+When a DLQ table is attached to a source, no errors are emitted to the source's
+*errs stream* (see the [error handling section][error-handling-docs] of the
+`compute::render` module).
+
+**Open questions:**
+
+  * Is there a better name for the option than `REDIRECT ERRORS`?
+  * What type gets reported in the system catalog for DLQ tables? Is it really a
+    table? Or is it a source or a subsource?
+  * Is it useful to report the `key` and the `value` in the DLQ relation? The
+    raw bytes will not be human readable for binary formats like Avro, and even
+    for textual formats like JSON you'll need to cast the values to `text`
+    to make them human readable.
+  * Can the `REDIRECT ERRORS` option be added to an existing source (i.e., via
+    `ALTER SOURCE`)? If so, what is the behavior? Does it pick up for all new
+    messages?
+  * Does the `mz_timestamp` column make sense? Is there something better
+    to call this column? Do the semantics make sense? (Including an
+    `ingested_at` wall clock timestamp would be nice, but that wouldn't
+    be a deterministic function of reclocking.)
+
+#### Upsert errors
+
+While the DLQ is useful for alerting users to data quality issues, it doesn't
+help users answer the question "does my upsert source currently have any keys
+whose most recent value failed to decode?"
+
+To solve this problem, we propose the addition of an `INLINE ERRORS` option
+to `ENVELOPE UPSERT`:
+
+```sql
+CREATE SOURCE src TO KAFKA CONNECTION kconn ...
+ENVELOPE UPSERT (
+    VALUE DECODING ERRORS = ({INLINE | PROPAGATE}, ...)
+)
+```
+
+The default behavior, which matches today's behavior, is `PROPAGATE`. Value
+decoding errors are propagated to the DLQ table or the source's *err stream*,
+whichever is active.
+
+When the `INLINE` behavior is specified, the source's relation gains a nullable
+column named `error` with a type of `record(description: text, code: text)`. If
+the most recent value for a key has been successfully decoded, this column will
+be `NULL`. If the most recent value for a key was not succesfully decoded, this
+column will contain details about the error. Additionally, the value is forced
+into a single nullable column named `value` with a type reflecting the
+configured value format—flattening of record-type values into top-level columns
+does not occur.
+
+When the `PROPAGATE` behavior is not specified, value decoding errors are
+essentially *not* treated as errors. They are neither forwarded to the DLQ table
+nor the source *errs stream*.
```

Comment:
Up above it says:
> The default behavior, which matches today's behavior, is `PROPAGATE`.

So how would you not specify this behavior? Do we need a 3rd option `NO_PROPAGATION` or similar?

Reply:
Or is this saying you can specify `VALUE DECODING ERRORS` with no option values?

Reply:
I was thinking you could specify either or both, but not neither:

```
VALUE DECODING ERRORS = (PROPAGATE)
VALUE DECODING ERRORS = (INLINE)
VALUE DECODING ERRORS = (PROPAGATE, INLINE)
```

Neither would result in silently discarding bad data without putting it _anywhere_, which seems like a good way to let users shoot themselves in the foot.

Reply:
Ohh I see -- I misunderstood the intention of this statement. We are saying that if `PROPAGATE` is not specified (but `INLINE` is and must be), the errors will only be inline and not be sent anywhere else.

Reply:
Yes, exactly!